### PR TITLE
linuxarm64: enable rkmpp and rkrga for portable builds

### PR DESCRIPTION
**Pull request by bjia56**
This PR enables rkmpp and rkrga for portable Linux arm64 builds of ffmpeg. Both libraries are built statically and linked into the resulting binary.

**Changes**

- Enable static rkmpp and rkrga for Linux arm64 portable builds
- Add patch to rkmpp to fix collisions with libvpx
- Add patch to rkrga to enable static lib build
- Modify rkrga lib lookup to use pkgconfig

**Issues**

None

### CODE REVIEW

File: builder/patches/rkmpp/0000-fix-vpx-collisions.patch [hunk 1]
```diff
@@ -0,0 +1,154 @@
+From 154a74b0a75eaecd5022cef3cb520ee949b578d9 Mon Sep 17 00:00:00 2001
+From: Brett Jia <dev.bjia56@gmail.com>
+Date: Wed, 19 Jun 2024 11:44:23 -0400
+Subject: [PATCH] namespace vars that conflict with vpx
```
Comment (nyanmisaka) [inline builder/patches/rkmpp/0000-fix-vpx-collisions.patch]: Perhaps submitting this upstream? https://github.com/rockchip-linux/mpp/issues/580
Maintaining mpp patches in jellyfin-ffmpeg repo is not ideal for me.

File: builder/patches/rkrga/0000-build-static.patch [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+diff --git a/meson.build b/meson.build
+index be2c2e9bb249..6f774754d628 100644
+--- a/meson.build
++++ b/meson.build
+@@ -30,12 +30,11 @@ librga_srcs = [
+ 
+ incdir = include_directories('include', 'im2d_api', 'core/hardware', 'core/utils', 'core/3rdparty/libdrm/include/drm')
+ 
+-librga = shared_library(
++librga = static_library(
```
Comment (nyanmisaka) [inline builder/patches/rkrga/0000-build-static.patch]: `sed -i 's/shared_library/library/g' meson.build`

File: builder/scripts.d/60-rkmpp.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        -DCMAKE_INSTALL_PREFIX="$FFBUILD_PREFIX" \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DBUILD_SHARED_LIBS=OFF \
+        ..
+
+    make -j$(nproc)
+    make install
+
+    echo "Libs.private: -lstdc++" >> "$FFBUILD_PREFIX"/lib/pkgconfig/rockchip_mpp.pc
+}
```
Comment (nyanmisaka) [inline builder/scripts.d/60-rkmpp.sh]: I would prefer to write like this to match other scripts in the builder.

```
ffbuild_dockerbuild() {
    git-mini-clone "$SCRIPT_REPO" "$SCRIPT_COMMIT" rkmpp
    cd rkmpp

    mkdir rkmpp_build && cd rkmpp_build

    cmake -GNinja \
        -DCMAKE_TOOLCHAIN_FILE="$FFBUILD_CMAKE_TOOLCHAIN" \
        -DCMAKE_BUILD_TYPE=Release \
        -DCMAKE_INSTALL_PREFIX="$FFBUILD_PREFIX" \
        -DBUILD_TEST=OFF \
        -DBUILD_SHARED_LIBS=OFF ..

    ninja -j$(nproc)
    ninja install

    echo "Libs.private: -lstdc++" >> "$FFBUILD_PREFIX"/lib/pkgconfig/rockchip_mpp.pc
}
```

File: builder/scripts.d/60-rkrga.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        --default-library=static \
+        -Dcpp_args=-fpermissive \
+        -Dlibdrm=false \
+        -Dlibrga_demo=false
+
+    meson configure rkrga_build
+    ninja -C rkrga_build install
+
+    echo "Libs.private: -lstdc++" >> "$FFBUILD_PREFIX"/lib/pkgconfig/librga.pc
+}
```
Comment (nyanmisaka) [inline builder/scripts.d/60-rkrga.sh]: Ditto.

```
ffbuild_dockerbuild() {
    git-mini-clone "$SCRIPT_REPO" "$SCRIPT_COMMIT" rkrga
    cd rkrga

    sed -i 's/shared_library/library/g' meson.build

    mkdir rkrga_build && cd rkrga_build

    local myconf=(
        --prefix="$FFBUILD_PREFIX"
        --buildtype=release
        --default-library=static
        -Dcpp_args=-fpermissive
        -Dlibdrm=false
        -Dlibrga_demo=false
    )

    if [[ $TARGET == linux* ]]; then
        myconf+=(
            --cross-file=/cross.meson
        )
    else
        echo "Unknown target"
        return -1
    fi

    meson "${myconf[@]}" ..

    ninja -j$(nproc)
    ninja install

    echo "Libs.private: -lstdc++" >> "$FFBUILD_PREFIX"/lib/pkgconfig/librga.pc
}
```

File: debian/patches/0058-add-full-hwa-pipeline-for-rockchip-rk3588-platform.patch [hunk 5]
```diff
@@ -140,8 +140,8 @@ Index: jellyfin-ffmpeg/configure
                                 { enabled libdrm ||
                                   die "ERROR: rkmpp requires --enable-libdrm"; }
                               }
-+enabled rkrga             && require rkrga rga/RgaApi.h c_RkRgaBlit -lrga &&
-+                             { require rkrga rga/im2d.h querystring -lrga &&
++enabled rkrga             && require_pkg_config rkrga librga rga/RgaApi.h c_RkRgaBlit&&
```
Comment (nyanmisaka) [inline debian/patches/0058-add-full-hwa-pipeline-for-rockchip-rk3588-platform.patch]:
```suggestion
+enabled rkrga             && require_pkg_config rkrga librga rga/RgaApi.h c_RkRgaBlit &&
```

**Comment by nyanmisaka** [comment]
60-rkmpp.sh -> 50-rkmpp.sh
60-rkrga.sh -> 50-rkrga.sh

50 or even lower should be enough, since they don't depend on libdrm or other libs.

**Comment by bjia56** [comment]
Comments addressed. I opened https://github.com/rockchip-linux/mpp/pull/621 for the upstream fix - the patch has been kept in the PR for the time being, in case there are any blockers in upstreaming it.

**Review by nyanmisaka — APPROVED** [review]

**Pull request merged**